doc/how-to-analyze-haskell-code-performance.rst

```diff
@@ -0,0 +1,126 @@
+How to analyze Haskell performance
+==================================
+
+When a Haskell application is slow or uses too much memory,
+Cabal and `GHC <https://downloads.haskell.org/ghc/latest/docs/users_guide/>`__
+can help you understand why.
+
+The main steps are to
+
+1. let GHC insert performance measuring code into your application,
+2. run the application to collect a performance report and
+3. visualize that report.
+
+The process of inserting performance measuring code and collecting performance information
+is called "profiling". The main work is done by GHC while Cabal only acts
+as a build configuration interface and passes corresponding compiler flags to GHC.
+
+Profiling CPU performance
+-------------------------
+
+The first step to build your application, e.g. ``my-app``, with profiling enabled, and
+the second step to run it to collect a report, can be done with a single ``cabal run`` command:
+
+.. code-block:: console
+
+      $ cabal run --enable-profiling --profiling-detail=late my-app -- +RTS -pj -RTS
+      <program runs and finishes>
+
+Finally, a profiling JSON report is written to a ``<app-name>.prof`` file,
+i.e. ``my-app.prof``, in the current directory.
+Load the profiling report file  ``my-app.prof`` into a visualizer
+and look for performance bottlenecks. One popular open-source
+`flame graph <https://www.brendangregg.com/flamegraphs.html>`__
+visualizer is
+`Speedscope <https://speedscope.app>`__,
+which runs in the browser and can open this JSON file directly.
```

Comment:
I would mention something like:

- `-pj` produces JSON output, can be visualized in `speedscope`.
- `-p` produces GHC's own `.prof` format, can be visualized in `profiteur` or `ghcprofview`.
- `-l -p` produces an eventlog that can be visualized in `speedscope` by first converting it via `hs-speedscope`.

Reply:
Also I'm not 100% sure of that I'm about to say, but my understanding is that `-p` and `-pj` shows total time spent, and even in speedscope you see totals. One can see how much time a function took, but not when.

I think the eventlog path shows totals but also shows when this happened, i.e. information will be interleaved with GCs or switching contexts, and split by capabilities. IMHO this is more useful and the thing I usually use.

Reply:
Added the three examples, although i think you might have meant `-l-p`, because `-l -p` (with a space inbetween) just produces two separate files `.eventlog` and `.prof`.

Reply:
Added a comment in my last review. It is actually `-l -p`. `-l-p` is "eventlog without parallel sparks" which is not what we want here.

Reply:
Closing this comment in favor of the other one.